factoryWithTypeCheckers.js

```diff
@@ -270,6 +272,74 @@ module.exports = function(isValidElement, throwOnDirectAccess) {
     return createChainableTypeChecker(validate);
   }
 
+  function createMapOfTypeChecker(typeCheckers) {
+    function validate(props, propName, componentName, location, propFullName) {
+      if (!Array.isArray(typeCheckers)) {
+        return new PropTypeError('Correct syntax is `PropTypes.mapOf([PropTypes.keyType, PropTypes.valueType])`.');
+      }
+      var keyChecker = typeCheckers[0];
+      var valueChecker = typeCheckers[1];
+      if (typeof Map !== 'function') {
+        return new PropTypeError('Cannot use `PropTypes.mapOf` if the Map data structure is not present');
+      }
+      if (typeof keyChecker !== 'function' || typeof valueChecker !== 'function') {
+        return new PropTypeError('Property `' + propFullName + '` of component `' + componentName + '` has invalid PropType notation inside mapOf.');
+      }
+      var propValue = props[propName];
+      if (!(propValue instanceof Map)) {
```

Comment:
`instanceof` is not actually reliable - it doesn't work cross-realm, and it can be forged (Error can't be brand checked yet, so `instanceof Error` is the best that can be done). To be robust here, we'd have to add a brand check - one complex enough that we'd want to add a dependency (one that i'm in the process of creating anyways, for another use case) that would potentially add to the bundle size.

Reply:
Did you see [my comment](https://github.com/facebook/prop-types/issues/190#issuecomment-526350505) about that? Also I'm curious to know more about the dependency you're creating. I'd sort of spitballed something of the sort a while back: https://github.com/conartist6/structure-ish. Its aim was never to replace an `instanceof` check though in the sense that it wouldn't detect any cross-realm set, only those that declared themselves in a compatible fashion.

Reply:
Regardless of structure-ish though I though the solutioning I proposed was pretty doable. Would you like me to implement it?

Reply:
If you're curious about how I would go about really solving the cross-realm problem, my answer is in this issue: https://github.com/conartist6/structure-ish/issues/1. Note that this solution would also support Immutable.js data types. I had put this project on the backburner while awaiting feedback on my proposal to integrate with Immutable, but immutable hasn't been actively maintained for a long time now, so -- so much for that.

Reply:
There's no need for that; the mechanisms are built into the language. I'll post links to the packages once they're published.

Reply:
And which mechanisms are they? You're referring to your technique of calling the size getter on a potential set? If you do that will you also be sure to also check for the existence of an Iterator?